Readme.md

# Swipe

  Swipe component with touch support, for image carousels or any other content. Try it out the [demo](http://component.github.com/swipe/) in your browser or on your device.

## Installation

    $ component install component/swipe

## Events

- `showing` (i, pane): when the transition to another pane begins
- `show` (i, pane): when the transition to another pane finishes

## API

### Swipe(el)

  Create a swipe object for `el`. This should be a container element
  that wraps a list of several items. View ./example.html for a
  working example.

### .threshold(n)

  Set the swipe threshold to `n`.

  This is the factor required for swipe
  to detect when a slide has passed the
  given threshold, and may display the next
  or previous slide. For example the default
  of `.5` means that the user must swipe _beyond_
  half of the side width.

### .fastThreshold(ms)

 Set the "fast" swipe threshold to `ms`.

 This is the amount of time in milliseconds
 which determines if a swipe was "fast" or not. When
 the swipe's duration is less than `ms` only 1/10th of
 the slide's width must be exceeded to display the previous
 or next slide.

### .duration(ms)

  Set the transition duration, defaults to 300ms.

### .interval(ms)

  Set the cycle interval, defaults to 5000ms.

### .refresh()

  This method should be invoked when the swipe element
  has been resized, or an item has been added or removed.

### .play()

  Play through all items using the cycle interval.

### .stop()

  Stop playing.

### .isFirst()

  Is on the first visible item.

### .isLast()

  Is on the last visisble item.

### .prev(cycle)

  Show the previous item.
  By default, do nothing if showing the first item.
  If `cycle` is true and we are on the first item, display the last item.

### .next(cycle)

  Show the next item.
  By default, do nothing if showing the last item.
  If `cycle` is true and we are on the last item, display the first item.

### .show(i, [ms], [options])

  Show item with the given index `i` with the given
  transition in `ms` defaulting to the `.duration()` value.

  You may pass `{ silent: true }` as an option to silence show events.

## License

  MIT


History.md

1.7.7 / 2015-04-15
==================

 * index: fix require of component/events

1.7.6 / 2015-03-16
==================

  * package: update all deps and fix browserify build
  * package: remove "development" field
  * package: remove "repo" field
  * component: update "emitter" to v1.2.0

1.7.5 / 2014-09-27
==================

  * component: update dependencies (#28)

1.7.4 / 2013-11-15
==================

  * Revert "use component/indexof"

1.7.3 / 2013-11-13
==================

  * package: use GitHub tarball URL for indexof component

1.7.2 / 2013-11-13
==================

  * Add touch support for IE (@tootallnate)
  * use component/indexof (@tootallnate)

1.7.1 / 2013-07-31
==================

  * fixed jumpiness when scrolling and quickly stopping scroll

1.7.0 / 2013-07-11
==================

  * replace `show` event with `showing`, wait till transitionend to emit `show`

1.6.0 / 2013-06-19
==================

  * add Swipe#fastThreshold(ms)
  * add Swipe#threshold(n)

1.5.4 / 2013-06-01
==================

  * index: fix `silent` value passed

1.5.3 / 2013-06-01
==================

  * index: make `refresh` not emit events

1.5.2 / 2013-05-31
==================

  * fix show on .refresh() regression
  * fix support for hidden slides
  * fix .isFirst(), should be visible only
  * change currentIndex -> currentVisibleIndex

1.5.1 / 2013-05-29
==================

  * fix .current relative to all elements and a few other subtle bugs

1.5.0 / 2013-05-28
==================

  * swipe is now display:none aware [MatthewMueller]

1.4.2 / 2013-05-27
==================

  * pin deps

1.4.1 / 2013-05-03
==================

  * remove .add()
  * remove .remove()
  * change .refresh() to detect dom changes

1.4.0 / 2013-05-02
==================

  * add `.add(el, i)`
  * add `.remove(i)`

1.3.4 / 2013-04-17
==================

  * swipe directionally aware

1.3.3 / 2013-04-05
==================

  * index: implemented proper feature detection

1.3.2 / 2013-03-27
==================

  * index: make sure to give extra room for zoom changes

1.3.1 / 2013-03-24
==================

  * index: prevent bad `show` if `refresh` is called right after `cycle`
  * index: make refresh `show` silent

1.3.0 / 2013-03-05
==================

  * add .show() .silent option

1.2.2 / 2013-02-26
==================

  * fix .refresh() - show the current item again

1.2.1 / 2013-02-15
==================

  * typo

1.2.0 / 2013-02-15
==================

  * add .refresh()

1.1.1 / 2013-01-18
==================

  * Add \`show\` event emission.

1.1.0 / 2012-12-16
==================

  * rename to swipe
  * fix clicking of last item

1.0.1 / 2012-11-22
==================

  * fix resistance. Closes #3

1.0.0 / 2012-11-22
==================

  * Initial release


build.js
/**
 * Require the given path.
 *
 * @param {String} path
 * @return {Object} exports
 * @api public
 */

function require(p, parent, orig){
  var path = require.resolve(p)
    , mod = require.modules[path];

  // lookup failed
  if (null == path) {
    orig = orig || p;
    parent = parent || 'root';
    throw new Error('failed to require "' + orig + '" from "' + parent + '"');
  }

  // perform real require()
  // by invoking the module's
  // registered function
  if (!mod.exports) {
    mod.exports = {};
    mod.client = mod.component = true;
    mod.call(this, mod, mod.exports, require.relative(path));
  }

  return mod.exports;
}

/**
 * Registered modules.
 */

require.modules = {};

/**
 * Registered aliases.
 */

require.aliases = {};

/**
 * Resolve `path`.
 *
 * Lookup:
 *
 *   - PATH/index.js
 *   - PATH.js
 *   - PATH
 *
 * @param {String} path
 * @return {String} path or null
 * @api private
 */

require.resolve = function(path){
  var orig = path
    , reg = path + '.js'
    , regJSON = path + '.json'
    , index = path + '/index.js'
    , indexJSON = path + '/index.json';

  return require.modules[reg] && reg
    || require.modules[regJSON] && regJSON
    || require.modules[index] && index
    || require.modules[indexJSON] && indexJSON
    || require.modules[orig] && orig
    || require.aliases[index];
};

/**
 * Normalize `path` relative to the current path.
 *
 * @param {String} curr
 * @param {String} path
 * @return {String}
 * @api private
 */

require.normalize = function(curr, path) {
  var segs = [];

  if ('.' != path.charAt(0)) return path;

  curr = curr.split('/');
  path = path.split('/');

  for (var i = 0; i < path.length; ++i) {
    if ('..' == path[i]) {
      curr.pop();
    } else if ('.' != path[i] && '' != path[i]) {
      segs.push(path[i]);
    }
  }

  return curr.concat(segs).join('/');
};

/**
 * Register module at `path` with callback `fn`.
 *
 * @param {String} path
 * @param {Function} fn
 * @api private
 */

require.register = function(path, fn){
  require.modules[path] = fn;
};

/**
 * Alias a module definition.
 *
 * @param {String} from
 * @param {String} to
 * @api private
 */

require.alias = function(from, to){
  var fn = require.modules[from];
  if (!fn) throw new Error('failed to alias "' + from + '", it does not exist');
  require.aliases[to] = from;
};

/**
 * Return a require function relative to the `parent` path.
 *
 * @param {String} parent
 * @return {Function}
 * @api private
 */

require.relative = function(parent) {
  var p = require.normalize(parent, '..');

  /**
   * lastIndexOf helper.
   */

  function lastIndexOf(arr, obj){
    var i = arr.length;
    while (i--) {
      if (arr[i] === obj) return i;
    }
    return -1;
  }

  /**
   * The relative require() itself.
   */

  function fn(path){
    var orig = path;
    path = fn.resolve(path);
    return require(path, parent, orig);
  }

  /**
   * Resolve relative to the parent.
   */

  fn.resolve = function(path){
    // resolve deps by returning
    // the dep in the nearest "deps"
    // directory
    if ('.' != path.charAt(0)) {
      var segs = parent.split('/');
      var i = lastIndexOf(segs, 'deps') + 1;
      if (!i) i = 0;
      path = segs.slice(0, i + 1).join('/') + '/deps/' + path;
      return path;
    }
    return require.normalize(p, path);
  };

  /**
   * Check if module is defined at `path`.
   */

  fn.exists = function(path){
    return !! require.modules[fn.resolve(path)];
  };

  return fn;
};require.register("component-event/index.js", function(module, exports, require){

/**
 * Bind `el` event `type` to `fn`.
 *
 * @param {Element} el
 * @param {String} type
 * @param {Function} fn
 * @param {Boolean} capture
 * @api public
 */

exports.bind = function(el, type, fn, capture){
  if (el.addEventListener) {
    el.addEventListener(type, fn, capture);
  } else {
    el.attachEvent('on' + type, fn);
  }
};

/**
 * Unbind `el` event `type`'s callback `fn`.
 *
 * @param {Element} el
 * @param {String} type
 * @param {Function} fn
 * @param {Boolean} capture
 * @api public
 */

exports.unbind = function(el, type, fn, capture){
  if (el.removeEventListener) {
    el.removeEventListener(type, fn, capture);
  } else {
    el.detachEvent('on' + type, fn);
  }
};
});
require.register("carousel/index.js", function(module, exports, require){

/**
 * Module dependencies.
 */

var event = require('event');

/**
 * Expose `Carousel`.
 */

module.exports = Carousel;

/**
 * Turn `el` into a carousel.
 *
 * @param {Element} el
 * @api public
 */

function Carousel(el) {
  if (!(this instanceof Carousel)) return new Carousel(el);
  if (!el) throw new TypeError('swipe() requires an element');
  this.el = el;
  this.child = el.children[0];
  this.total = this.child.children.length;
  this.childWidth = el.getBoundingClientRect().width;
  this.width = this.childWidth * this.total | 0;
  this.child.style.width = this.width + 'px';
  this.child.style.height = this.height + 'px';
  this.interval(5000);
  this.duration(300);
  this.show(0, 0);
  this.bind();
  // TODO: Swipeable() mixin
}

/**
 * Bind event handlers.
 *
 * @api public
 */

Carousel.prototype.bind = function(){
  // TODO: create a bulk event management component.. this is nasty
  event.bind(this.child, 'mousedown', this.ontouchstart.bind(this));
  event.bind(this.child, 'mousemove', this.ontouchmove.bind(this));
  event.bind(document, 'mouseup', this.ontouchend.bind(this));
  event.bind(this.child, 'touchstart', this.ontouchstart.bind(this));
  event.bind(this.child, 'touchmove', this.ontouchmove.bind(this));
  event.bind(document, 'touchend', this.ontouchend.bind(this));
};

/**
 * Unbind event handlers.
 *
 * @api public
 */

Carousel.prototype.unbind = function(){
  // TODO: me
};

/**
 * Handle touchstart.
 *
 * @api private
 */

Carousel.prototype.ontouchstart = function(e){
  e.stopPropagation();
  if (e.touches) e = e.touches[0];

  this.transitionDuration(0);
  this.dx = 0;

  this.down = {
    x: e.pageX,
    at: new Date
  };
};

/**
 * Handle touchmove.
 *
 * For the first and last slides
 * we apply some resistence to help
 * indicate that you're at the edges.
 *
 * @api private
 */

Carousel.prototype.ontouchmove = function(e){
  if (!this.down) return;
  if (e.touches && e.touches.length > 1) return;
  e.stopPropagation();
  e.preventDefault();
  if (e.touches) e = e.touches[0];
  var s = this.down;
  var x = e.pageX;
  var w = this.childWidth;
  var i = this.current;
  this.dx = x - s.x;
  var dir = this.dx < 0 ? 1 : 0;
  if (this.isFirst() && 0 == dir) this.dx /= 2;
  if (this.isLast() && 1 == dir) this.dx /= 2;
  this.translate((i * w) + -this.dx);
};

/**
 * Handle touchend.
 *
 * @api private
 */

Carousel.prototype.ontouchend = function(e){
  if (!this.down) return;
  e.stopPropagation();

  // touches
  if (e.changedTouches) e = e.changedTouches[0];

  // setup
  var dx = this.dx;
  var x = e.pageX;
  var w = this.childWidth;

  // < 200ms swipe
  var ms = new Date - this.down.at;
  var threshold = ms < 200 ? w / 10 : w / 2;
  var dir = dx < 0 ? 1 : 0;
  var half = Math.abs(dx) >= threshold;

  // clear
  this.down = null;

  // first -> next
  if (this.isFirst() && 1 == dir && half) return this.next();

  // first -> first
  if (this.isFirst()) return this.prev();

  // last -> last
  if (this.isLast() && 1 == dir) return this.next();

  // last -> prev
  if (this.isLast()) return this.prev();

  // N -> N + 1
  if (1 == dir && half) return this.next();

  // N -> N - 1
  if (0 == dir && half) return this.prev();

  // N -> N
  this.show(this.current);
};

/**
 * Set transition duration to `ms`.
 *
 * @param {Number} ms
 * @return {Carousel} self
 * @api public
 */

Carousel.prototype.duration = function(ms){
  this._duration = ms;
  return this;
};

/**
 * Set cycle interval to `ms`.
 *
 * @param {Number} ms
 * @return {Carousel} self
 * @api public
 */

Carousel.prototype.interval = function(ms){
  this._interval = ms;
  return this;
};

/**
 * Play through all the elements.
 *
 * @return {Carousel} self
 * @api public
 */

Carousel.prototype.play = function(){
  if (this.timer) return;
  this.timer = setInterval(this.cycle.bind(this), this._interval);
  return this;
};

/**
 * Stop playing.
 *
 * @return {Carousel} self
 * @api public
 */

Carousel.prototype.stop = function(){
  clearInterval(this.timer);
  return this;
};

/**
 * Show the next slide, when the end
 * is reached start from the beginning.
 *
 * @api public
 */

Carousel.prototype.cycle = function(){
  if (this.isLast()) {
    this.current = -1;
    this.next();
  } else {
    this.next();
  }
};

/**
 * Check if we're on the first slide.
 *
 * @return {Boolean}
 * @api public
 */

Carousel.prototype.isFirst = function(){
  return this.current == 0;
};

/**
 * Check if we're on the last slide.
 *
 * @return {Boolean}
 * @api public
 */

Carousel.prototype.isLast = function(){
  return this.current == this.total - 1;
};

/**
 * Show the previous slide, if any.
 *
 * @return {Carousel} self
 * @api public
 */

Carousel.prototype.prev = function(){
  this.show(this.current - 1);
  return this;
};

/**
 * Show the next slide, if any.
 *
 * @return {Carousel} self
 * @api public
 */

Carousel.prototype.next = function(){
  this.show(this.current + 1);
  return this;
};

/**
 * Show slide `i`.
 *
 * @param {Number} i
 * @return {Carousel} self
 * @api public
 */

Carousel.prototype.show = function(i, ms){
  if (null == ms) ms = this._duration;
  i = Math.max(0, Math.min(i, this.total - 1));
  var x = this.childWidth * i;
  this.current = i;
  this.transitionDuration(ms);
  this.translate(x);
  return this;
};

/**
 * Set transition duration.
 *
 * @api private
 */

Carousel.prototype.transitionDuration = function(ms){
  var s = this.child.style;
  s.webkitTransitionDuration =
  s.MozTransitionDuration =
  s.msTransitionDuration =
  s.OTransitionDuration =
  s.transitionDuration = ms + 'ms';
};

/**
 * Translate to `x`.
 *
 * @api private
 */

Carousel.prototype.translate = function(x){
  var s = this.child.style;
  x = -x;
  s.webkitTransform = s.MozTransform = 'translate3d(' + x + 'px, 0, 0)';
  s.msTransform = s.OTransform = 'translateX(' + x + 'px)';
};


});
require.alias("component-event/index.js", "carousel/deps/event/index.js");


index.js

/**
 * Module dependencies.
 */

var transitionend = require('transitionend-property');
var transform = require('transform-property');
var touchAction = require('touchaction-property');
var has3d = require('has-translate3d');
var style = require('computed-style');
var Emitter = require('emitter');
var event = require('event');
var events = require('component-events');
var min = Math.min;
var max = Math.max;

/**
 * Expose `Swipe`.
 */

module.exports = Swipe;

/**
 * Turn `el` into a swipeable list.
 *
 * @param {Element} el
 * @api public
 */

function Swipe(el) {
  if (!(this instanceof Swipe)) return new Swipe(el);
  if (!el) throw new TypeError('Swipe() requires an element');
  this.child = el.children[0];
  this.touchAction('none');
  this.currentEl = this.children().visible[0];
  this.currentVisible = 0;
  this.current = 0;
  this.el = el;
  this.refresh();
  this.interval(5000);
  this.duration(300);
  this.fastThreshold(200);
  this.threshold(0.5);
  this.show(0, 0, { silent: true });
  this.bind();
}

/**
 * Mixin `Emitter`.
 */

Emitter(Swipe.prototype);

/**
 * Set the swipe threshold to `n`.
 *
 * This is the factor required for swipe
 * to detect when a slide has passed the
 * given threshold, and may display the next
 * or previous slide. For example the default
 * of `.5` means that the user must swipe _beyond_
 * half of the side width.
 *
 * @param {Number} n
 * @api public
 */

Swipe.prototype.threshold = function(n){
  this._threshold = n;
};

/**
 * Set the "fast" swipe threshold to `ms`.
 *
 * This is the amount of time in milliseconds
 * which determines if a swipe was "fast" or not. When
 * the swipe's duration is less than `ms` only 1/10th of
 * the slide's width must be exceeded to display the previous
 * or next slide.
 *
 * @param {Number} n
 * @api public
 */

Swipe.prototype.fastThreshold = function(ms){
  this._fastThreshold = ms;
};

/**
 * Refresh sizing data.
 *
 * @api public
 */

Swipe.prototype.refresh = function(){
  var children = this.children();
  var visible = children.visible.length;
  var prev = this.visible || visible;

  var i = indexOf(children.visible, this.currentEl);

  // we removed/added item(s), update current
  if (visible < prev && i <= this.currentVisible && i >= 0) {
    this.currentVisible -= this.currentVisible - i;
  } else if (visible > prev && i > this.currentVisible) {
    this.currentVisible += i - this.currentVisible;
  }

  this.visible = visible;
  this.childWidth = [0];
  this.width = 0;
  for(var i = 0, size = this.children().all.length; i < size; i++) {
    this.width += this.children().all[i].clientWidth;
    this.childWidth.push(this.width);
  }
  this.child.style.width = this.width + 'px';
  this.child.style.height = this.height + 'px';
  this.show(this.currentVisible, 0, { silent: true });
};

/**
 * Bind event handlers.
 *
 * @api public
 */

Swipe.prototype.bind = function(){
  this.events = events(this.child, this);
  this.docEvents = events(document, this);

  // standard mouse click events
  this.events.bind('mousedown', 'ontouchstart');
  this.events.bind('mousemove', 'ontouchmove');
  this.docEvents.bind('mouseup', 'ontouchend');

  // W3C touch events
  this.events.bind('touchstart');
  this.events.bind('touchmove');
  this.docEvents.bind('touchend');

  // MS IE touch events
  this.events.bind('PointerDown', 'ontouchstart');
  this.events.bind('PointerMove', 'ontouchmove');
  this.docEvents.bind('PointerUp', 'ontouchstart');
};

/**
 * Unbind event handlers.
 *
 * @api public
 */

Swipe.prototype.unbind = function(){
  this.events.unbind();
  this.docEvents.unbind();
};

/**
 * Handle touchstart.
 *
 * @api private
 */

Swipe.prototype.ontouchstart = function(e){
  this.transitionDuration(0);
  this.dx = 0;
  this.updown = null;

  var touch = this.getTouch(e);
  this.down = {
    x: touch.pageX,
    y: touch.pageY,
    at: new Date()
  };
};

/**
 * Handle touchmove.
 *
 * For the first and last slides
 * we apply some resistence to help
 * indicate that you're at the edges.
 *
 * @api private
 */

Swipe.prototype.ontouchmove = function(e){
  if (!this.down || this.updown) return;
  var touch = this.getTouch(e);

  // TODO: ignore more than one finger
  if (!touch) return;

  var down = this.down;
  var x = touch.pageX;
  var w = this.childWidth;
  var i = this.currentVisible;
  this.dx = x - down.x;

  // determine dy and the slope
  if (null == this.updown) {
    var y = touch.pageY;
    var dy = y - down.y;
    var slope = dy / this.dx;

    // if is greater than 1 or -1, we're swiping up/down
    if (slope > 1 || slope < -1) {
      this.updown = true;
      return;
    } else {
      this.updown = false;
    }
  }

  e.preventDefault();

  var dir = this.dx < 0 ? 1 : 0;
  if (this.isFirst() && 0 == dir) this.dx /= 2;
  if (this.isLast() && 1 == dir) this.dx /= 2;
  this.translate((i * w) + -this.dx);
};

/**
 * Handle touchend.
 *
 * @api private
 */

Swipe.prototype.ontouchend = function(e){
  e.stopPropagation();
  if (!this.down) return;
  var touch = this.getTouch(e);

  // setup
  var dx = this.dx;
  var x = touch.pageX;
  var w = this.childWidth;

  // < 200ms swipe
  var ms = new Date() - this.down.at;
  var threshold = ms < this._fastThreshold ? w / 10 : w * this._threshold;
  var dir = dx < 0 ? 1 : 0;
  var half = Math.abs(dx) >= threshold;

  // clear
  this.down = null;

  // first -> next
  if (this.isFirst() && 1 == dir && half) return this.next();

  // first -> first
  if (this.isFirst()) return this.prev();

  // last -> last
  if (this.isLast() && 1 == dir) return this.next();

  // N -> N + 1
  if (1 == dir && half) return this.next();

  // N -> N - 1
  if (0 == dir && half) return this.prev();

  // N -> N
  this.show(this.currentVisible);
};

/**
 * Set transition duration to `ms`.
 *
 * @param {Number} ms
 * @return {Swipe} self
 * @api public
 */

Swipe.prototype.duration = function(ms){
  this._duration = ms;
  return this;
};

/**
 * Set cycle interval to `ms`.
 *
 * @param {Number} ms
 * @return {Swipe} self
 * @api public
 */

Swipe.prototype.interval = function(ms){
  this._interval = ms;
  return this;
};

/**
 * Play through all the elements.
 *
 * @return {Swipe} self
 * @api public
 */

Swipe.prototype.play = function(){
  if (this.timer) return;
  this.timer = setInterval(this.cycle.bind(this), this._interval);
  return this;
};

/**
 * Stop playing.
 *
 * @return {Swipe} self
 * @api public
 */

Swipe.prototype.stop = function(){
  clearInterval(this.timer);
  this.timer = null;
  return this;
};

/**
 * Show the next slide, when the end
 * is reached start from the beginning.
 *
 * @api public
 */

Swipe.prototype.cycle = function(){
  if (this.isLast()) {
    this.currentVisible = -1;
    this.next();
  } else {
    this.next();
  }
};

/**
 * Check if we're on the first visible slide.
 *
 * @return {Boolean}
 * @api public
 */

Swipe.prototype.isFirst = function(){
  return this.currentVisible == 0;
};

/**
 * Check if we're on the last visible slide.
 *
 * @return {Boolean}
 * @api public
 */

Swipe.prototype.isLast = function(){
  return this.currentVisible == this.visible - 1;
};

/**
 * Show the previous slide, if any.
 *
 * @return {Swipe} self
 * @api public
 */

Swipe.prototype.prev = function(){
  this.show(this.currentVisible - 1);
  return this;
};

/**
 * Show the next slide, if any.
 *
 * @return {Swipe} self
 * @api public
 */

Swipe.prototype.next = function(){
  this.show(this.currentVisible + 1);
  return this;
};

/**
 * Show slide `i`.
 *
 * Emits `show `event
 *
 * @param {Number} i
 * @return {Swipe} self
 * @api public
 */

Swipe.prototype.show = function(i, ms, options){
  options = options || {};
  if (null == ms) ms = this._duration;
  var self = this;
  var children = this.children();
  i = max(0, min(i, children.visible.length - 1));
  this.currentVisible = i;
  this.currentEl = children.visible[i];
  this.current = indexOf(children.all, this.currentEl);
  this.transitionDuration(ms);
  this.translate(this.childWidth[i]);

  if (!options.silent) {
    this.emit('showing', this.current, this.currentEl);
    if (!ms) return this;
    event.bind(this.child, transitionend, function shown() {
      if (self.current == i) self.emit('show', self.current, self.currentEl);
      event.unbind(self.child, transitionend, shown);
    });
  }
  return this;
};

/**
 * Return children categorized by visibility.
 *
 * @return {Object}
 * @api private
 */

Swipe.prototype.children = function(){
  var els = this.child.children;

  var ret = {
    all: els,
    visible: [],
    hidden: []
  };

  for (var i = 0; i < els.length; i++) {
    var el = els[i];
    if (visible(el)) {
      ret.visible.push(el);
    } else {
      ret.hidden.push(el);
    }
  }

  return ret;
};

/**
 * Set transition duration.
 *
 * @api private
 */

Swipe.prototype.transitionDuration = function(ms){
  var s = this.child.style;
  s.webkitTransition = ms + 'ms -webkit-transform';
  s.MozTransition = ms + 'ms -moz-transform';
  s.msTransition = ms + 'ms -ms-transform';
  s.OTransition = ms + 'ms -o-transform';
  s.transition = ms + 'ms transform';
};

/**
 * Translate to `x`.
 *
 * TODO: use translate component
 *
 * @api private
 */

Swipe.prototype.translate = function(x){
  var s = this.child.style;
  x = -x;
  if (has3d) {
    s[transform] = 'translate3d(' + x + 'px, 0, 0)';
  } else {
    s[transform] = 'translateX(' + x + 'px)';
  }
};

/**
 * Sets the "touchAction" CSS style property to `value`.
 *
 * @api private
 */

Swipe.prototype.touchAction = function(value){
  var s = this.child.style;
  if (touchAction) {
    s[touchAction] = value;
  }
};

/**
 * Gets the appropriate "touch" object for the `e` event. The event may be from
 * a "mouse", "touch", or "Pointer" event, so the normalization happens here.
 *
 * @api private
 */

Swipe.prototype.getTouch = function(e){
  // "mouse" and "Pointer" events just use the event object itself
  var touch = e;
  if (e.changedTouches && e.changedTouches.length > 0) {
    // W3C "touch" events use the `changedTouches` array
    touch = e.changedTouches[0];
  }
  return touch;
};

/**
 * Return index of `el` in `els`.
 *
 * @param {Array} els
 * @param {Element} el
 * @return {Number}
 * @api private
 */

function indexOf(els, el) {
  for (var i = 0; i < els.length; i++) {
    if (els[i] == el) return i;
  }
  return -1;
}

/**
 * Check if `el` is visible.
 *
 * @param {Element} el
 * @return {Boolean}
 * @api private
 */

function visible(el) {
  return style(el).display != 'none';
}
